ports/geckolib/values.rs

```diff
@@ -12,65 +12,147 @@ use style::values::computed::{LengthOrPercentage, LengthOrPercentageOrAuto, Leng
     }
 
+    #[inline]
     fn set_auto(&mut self) {
         self.mUnit = nsStyleUnit::eStyleUnit_Auto;
         unsafe { *self.mValue.mInt.as_mut() = 0; }
     }
+    #[inline]
+    fn is_auto(&self) -> bool {
+        self.mUnit == nsStyleUnit::eStyleUnit_Auto
+    }
 
+    #[inline]
     fn set_normal(&mut self) {
         self.mUnit = nsStyleUnit::eStyleUnit_Normal;
         unsafe { *self.mValue.mInt.as_mut() = 0; }
     }
+    #[inline]
+    fn is_normal(&self) -> bool {
+        self.mUnit == nsStyleUnit::eStyleUnit_Normal
+    }
 
+    #[inline]
     fn set_coord(&mut self, val: Au) {
         self.mUnit = nsStyleUnit::eStyleUnit_Coord;
         unsafe { *self.mValue.mInt.as_mut() = val.0; }
     }
-
-    fn set_percent(&mut self, val: f32) {
-        self.mUnit = nsStyleUnit::eStyleUnit_Percent;
-        unsafe { *self.mValue.mFloat.as_mut() = val; }
+    #[inline]
+    fn is_coord(&self) -> bool {
+        self.mUnit == nsStyleUnit::eStyleUnit_Coord
+    }
+    #[inline]
+    fn get_coord(&self) -> Au {
+        debug_assert!(self.is_coord());
+        Au(unsafe { *self.mValue.mInt.as_ref() })
     }
 
+    #[inline]
     fn set_int(&mut self, val: i32) {
         self.mUnit = nsStyleUnit::eStyleUnit_Integer;
         unsafe { *self.mValue.mInt.as_mut() = val; }
     }
+    #[inline]
+    fn is_int(&self) -> bool {
+        self.mUnit == nsStyleUnit::eStyleUnit_Integer
+    }
+    #[inline]
+    fn get_int(&self) -> i32 {
+        debug_assert!(self.is_int());
+        unsafe { *self.mValue.mInt.as_ref() }
+    }
 
+    #[inline]
     fn set_enum(&mut self, val: i32) {
         self.mUnit = nsStyleUnit::eStyleUnit_Enumerated;
         unsafe { *self.mValue.mInt.as_mut() = val; }
     }
+    #[inline]
+    fn is_enum(&self) -> bool {
+        self.mUnit == nsStyleUnit::eStyleUnit_Enumerated
+    }
+    #[inline]
+    fn get_enum(&self) -> i32 {
+        debug_assert!(self.is_enum());
+        unsafe { *self.mValue.mInt.as_ref() }
+    }
+
+    #[inline]
+    fn set_percent(&mut self, val: f32) {
+        self.mUnit = nsStyleUnit::eStyleUnit_Percent;
+        unsafe { *self.mValue.mFloat.as_mut() = val; }
+    }
+    #[inline]
+    fn is_percent(&self) -> bool {
+        self.mUnit == nsStyleUnit::eStyleUnit_Percent
+    }
+    #[inline]
+    fn get_percent(&self) -> f32 {
+        debug_assert!(self.is_percent());
+        unsafe { *self.mValue.mFloat.as_ref() }
+    }
 
+    #[inline]
     fn set_factor(&mut self, val: f32) {
         self.mUnit = nsStyleUnit::eStyleUnit_Factor;
         unsafe { *self.mValue.mFloat.as_mut() = val; }
     }
+    #[inline]
+    fn is_factor(&self) -> bool {
+        self.mUnit == nsStyleUnit::eStyleUnit_Factor
+    }
+    #[inline]
+    fn get_factor(&self) -> f32 {
+        debug_assert!(self.is_factor());
+        unsafe { *self.mValue.mFloat.as_ref() }
+    }
 }
 
-pub trait ToGeckoStyleCoord {
+pub trait ToGeckoStyleCoord : Sized {
```

Comment:
Might want to rename this to GeckoStyleCoordConversion, or some potentially-more-idiomatic name that reflects the bidirectional conversion.